File: Cuarto/Laboratorio/Talleres/RobocodeLabo/src/Robotito/EstrategiaEvasiva.java
package Robotito;

import robocode.JuniorRobot;

public class EstrategiaEvasiva implements CombatStrategy {

  private boolean ready = false;
  private final int distance = 100;
  private final int terminatorThreshold = 5;
  private final int leftWallAngle = 270;

  @Override
  public void init(JuniorRobot robot) {
    ready = robot.others < terminatorThreshold;

    robot.setColors(JuniorRobot.yellow, JuniorRobot.purple, JuniorRobot.purple,
        JuniorRobot.purple, JuniorRobot.red);

    robot.turnTo(leftWallAngle);
    robot.turnGunTo(0);
    robot.ahead(robot.fieldWidth);
  }

  @Override
  public void onTick(JuniorRobot robot) {
    robot.fire(CombatStrategy.bulletHalfPower);
    if (robot.others < terminatorThreshold) {
      ready = true;
    }
    robot.ahead(distance);
    robot.doNothing();
  }

  @Override
  public void onHitWall(JuniorRobot robot) {
    robot.turnRight(90);
    robot.ahead(distance);
    // hacia el INTERIOR del mapa
    int inward = (robot.heading + 90) % 360;
    robot.turnGunTo((inward) % 360);
  }

  @Override
  public void onScannedRobot(JuniorRobot robot) {

  }

  @Override
  public void onHitRobot(JuniorRobot robot) {

    robot.turnGunTo(robot.hitRobotAngle);
    robot.fire(CombatStrategy.bulletMaxPower);
  }

  @Override
  public void onHitByBullet(JuniorRobot robot) {

  }

  public boolean readyToAttack() {
    return ready;
  }
}

File: Cuarto/Laboratorio/Talleres/RobocodeLabo/src/RobotitoKotlin/EstrategiaEvasiva.kt
package RobotitoKotlin

import robocode.JuniorRobot


class EstrategiaEvasiva(robot: JuniorRobot) : CombatStrategy(robot) {
    val terminatorThreshold= 5
    val leftWallAngle= 270
    var ready= false
    val distance = 100
    init{
        ready = robot.others < this.terminatorThreshold;

        robot.setColors(
            JuniorRobot.yellow, JuniorRobot.purple, JuniorRobot.red,
            JuniorRobot.purple, JuniorRobot.red)

        robot.turnTo(leftWallAngle)
        robot.turnGunTo(0)
        robot.ahead(robot.fieldWidth)
    }

    override fun onTick() {
        robot.fire(bulletHalfPower)
        if (robot.others < this.terminatorThreshold) {
            ready = true
        }
        robot.ahead(distance)
        robot.doNothing()
    }

    override fun onScannedRobot() {

    }

    override fun onHitWall() {
        robot.turnRight(90)
        robot.ahead(distance)

        // hacia el INTERIOR del mapa
        val inward = (robot.heading + 90) % 360
        robot.turnGunTo((inward) % 360)
    }

    override fun onHitRobot() {
        robot.turnGunTo(robot.hitRobotAngle)
        robot.fire(bulletMaxPower)
    }

    override fun onHitByBullet() {
    }

    override fun readyToAttack(): Boolean {
        return ready
    }

}